README.md
# secrBVN
Bivariate Normal Home Range SECR

This package supports work on the effect of range elongation on spatially explicit capture-recapture (SECR) estimates of density (Efford, M. G. In press. Non-circular home ranges and the estimation of population density. Ecology)

The package may be installed from R by running:

install.packages("devtools")

devtools::install_github("MurrayEfford/secrBVN", build_vignettes = TRUE)


NEWS
Changes in version 1.3.0 (2022-10-14)

New features or changed behaviour of old functions
    o   runEllipsesim() may generate new random traps at each replicate
    o   simplot() passes dots arguments to points()
    
Other    
    o   removed cluster code (multithreading in secr.fit)    
    o   geoR and rgeos now suggested 

Changes in version 1.2.3 (2018-08-14)

New features or changed behaviour of old functions
    o   runEllipseSim() returns both extractfn(fit) and predict(fit) for each replicate
    o   runEllipseSim() new argument outfile (name of cluster log)
    o   clarified arguments of runEllipseSim() and simcapt.bvn() wrt hazard and detectfn
    o   plotpopn.bvn() new argument region for truncation of plotted ellipses
    o   anisotropic.fit()
    
Other    
    o   GitHub repo murrayefford/secrBVN

Changes in version 1.2.1 (2018-07-11)

New features or changed behaviour of old functions
    o   simpopn.bvn() argument s2xy may be a function, allowing individual heterogeneity
    o   runEllipseSim() argument sigmaX may be a function as for simpopn.bvn()
    o   rebuilt for R 3.5.0
    
Changes in version 1.2.0 (2016-09-30)

New features or changed behaviour of old functions
    o   simcapt.bvn() allows multi-catch detectors

R/utils.R
## random ellipse utility functions

inside.ellipse <- function (xy, centres, theta, s2xy, p = 0.95) {
    xy <- unlist(xy)
    xyc <- centres
    xyc[,] <- NA
    for (i in 1:nrow(centres)) {
        xyc[i,] <- xy - centres[i,]
        xyc[i,] <- rotate (xyc[i,,drop=F], -360/2/pi*theta[i])
    }
    inside <- function (xy) {
        sum(xy^2/s2xy) < circular.r(p)^2
    }
    any(apply(xyc,1,inside))
}

pmvn <- function (xy, centres, theta, s2xy) {
    xy <- unlist(xy)
    xyc <- centres
    xyc[,] <- NA
    maxd <- 1/(2 * pi * prod(s2xy)^0.5)
    d <- logical(nrow(centres))
    for (i in 1:nrow(centres)) {
        xyc[i,] <- xy - centres[i,]
        xyc[i,] <- rotate (xyc[i,,drop=F], -360/2/pi*theta[i])
        d[i] <- runif(1) < (mvtnorm::dmvnorm(xyc[i,], mean=c(0,0),
                                    sigma = matrix(c(s2xy[1],0,0,s2xy[2]), nrow = 2))/maxd)
    }
    any(d)
}

ellipse <- function (center, shape, radius, segments = 51) {
    ## based on ellipse function in 'car' package
    angles <- (0:segments) * 2 * pi/segments
    unit.circle <- cbind(cos(angles), sin(angles))
    Q <- chol(shape, pivot = TRUE)
    order <- order(attr(Q, "pivot"))
    ellipse <- t(center + radius * t(unit.circle %*% Q[, order]))
    colnames(ellipse) <- c("x", "y")
    ellipse
}

random.ellipses <- function(popn, plt = T, s2xy = c(0.5,5), radius = 2.45) {
    popn <- as.matrix(popn)
    N <- nrow(popn)
    cov <- matrix(c(s2xy[1],0,0,s2xy[2]), ncol = 2)
    ## random orientation
    theta <- runif (N, max=2 * pi)
    for (i in 1:N) {
        ellipsi <- ellipse(popn[i,], shape=cov, radius = radius)
        ellipsi <- rotate (ellipsi, theta[i]*360/2/pi, centrexy = popn[i,])
        polygon (ellipsi)
    }

    xy <- cbind (runif(1000, min=par()$usr[1], max=par()$usr[2]),
                 runif(1000, min=par()$usr[3], max=par()$usr[4]))
    points(xy)
    OK <- apply(xy, 1, inside.ellipse, centres=popn, theta=theta, s2xy=s2xy)
    OK <- apply(xy, 1, pmvn, centres=popn, theta=theta, s2xy=s2xy)
    points(xy[OK,], pch=16)
}




R/sumplot.R
simsum <- function (simlist, component = c("fit", "pred"), parm = "D", trueval = 4, 
                    compact = c('av.nCH', 'nvalid', 'RB', 'RSE', 'rRMSE', 'COV'), 
                    dec = 3, maxfactor = 10) {
    sumD <- function(x) {
        getfield <- function (y, field = 'estimate') {
            out <- y[parm, field]
            if (!is.null(maxfactor))
                out[abs(out) > (maxfactor * trueval)] <- NA
            out
        }
        Dval <- sapply(lapply(x, '[[', component), getfield, 'estimate')
        DSE  <- sapply(lapply(x, '[[', component), getfield, 'SE.estimate')
        Dlcl <- sapply(lapply(x, '[[', component), getfield, 'lcl')
        Ducl <- sapply(lapply(x, '[[', component), getfield, 'ucl')
        n <- sum(!is.na(Dval))
        RB <- (Dval-trueval)/trueval
        RSE <- DSE/Dval
        COV <- (trueval>=Dlcl) & (trueval<=Ducl)
        npop <- sapply(x, '[[', 'npop')
        nCH <- sapply(x, '[[', 'nCH')
        c(av.npop = mean(npop), 
          av.nCH = mean(nCH), 
          nvalid = n,
          av.parmhat = mean(Dval, na.rm = TRUE),
          md.parmhat = median(Dval, na.rm = TRUE), 
          sd.parmhat = sd(Dval, na.rm = TRUE),
          RB = mean(RB, na.rm = TRUE), 
          seRB = sd(Dval, na.rm = TRUE)/trueval/n^0.5,
          RSE = mean(RSE, na.rm = TRUE), 
          seRSE = sd(RSE, na.rm = TRUE)/n^0.5,
          rRMSE = sqrt(mean((Dval-trueval)^2, na.rm = TRUE)) / trueval,
          COV = mean(COV, na.rm = TRUE))
    }
    tidy <- function (x) {
        tmp <- sapply(x, sumD)
        colnames(tmp) <- names(x)
        tmp
    }
    component <- match.arg(component)
    outlist <- lapply(simlist, tidy)
    if (is.null(compact)) {
        out <- lapply(outlist,round, dec)
    }
    else {
        comp <- function (out) {
            round(t(out)[,compact], dec)
        }
        out <- lapply(outlist, comp)
    }
    out
}

simplot <- function (simlist, component = c("fit", "pred"), parm = "D", trueval = 4, 
                     xval = 1:4, xlim = c(0.7,4.3), ylim = c(-0.2,0.2),
                     legend = TRUE, pchi = c(21, 16, 22, 24), cexi = rep(1.2,4), ...) {
    plotone <- function (out, i) {
        segments(xval+offset[i], out['RB',]-2*out['seRB',],
                 xval+offset[i], out['RB',]+2*out['seRB',])
        points(xval+offset[i], out['RB',], pch = pchi[i], bg = 'white', cex = cexi[i], ...)
    }
    component <- match.arg(component)
    outlist <- simsum(simlist, component = component, parm = parm, 
                      trueval = trueval, compact = NULL)
    nscen <- length(simlist)
    offset <- seq(-(nscen-1)*0.025, (nscen-1)*0.025, 0.05)
    plot(0,0, xlim=xlim, ylim = ylim, xlab = 'Aspect ratio',
         ylab = 'Relative bias',axes = FALSE, type = 'n')
    mapply(plotone, outlist, 1:length(outlist))
    axis(1, at = 1:4)
    axis(2, las = 1)
    abline(h=0, lty=2)
    if (legend)
        legend ((par()$usr[2] - par()$usr[1]) * 0.1 + par()$usr[1],  
                par()$usr[4]*0.95, legend = names(simlist),
                pch = pchi, cex = par()$cex * 0.8, pt.cex = cexi, adj = 0)
    invisible(outlist)
}


R/anisotropic.R
anisodistfn <- function (xy1, xy2, mask) {
    if (missing(xy1)) return(character(0))
    if (!requireNamespace("geoR")) stop ("aniso requires geoR that is unavailable")
    xy1 <- as.matrix(xy1)
    xy2 <- as.matrix(xy2)
    miscparm <- attr(mask, 'miscparm')
    psiA <- miscparm[1]           # anisotropy angle; identity link
    psiR <- 1 + exp(miscparm[2])  # anisotropy ratio; log link
    aniso.xy1 <- geoR::coords.aniso(xy1, aniso.pars = c(psiA, psiR))
    aniso.xy2 <- geoR::coords.aniso(xy2, aniso.pars = c(psiA, psiR))
    secr::edist(aniso.xy1, aniso.xy2) # nrow(xy1) x nrow(xy2) matrix
}

predictAniso <- function (fit, angle = c("degrees", "radians")) {
    angle <- match.arg(angle)
    co <- coef(fit)
    # if (!all( c("psiA", "psiR")  %in% rownames(co)))
    #    stop ("input is not anisotropic.fit")
    pred <- co[c("psiA", "psiR"), ]
    rownames(pred) <- c("psiA", "psiR")
    colnames(pred)[1:2] <- c("estimate", "SE.estimate")
    if (!is.null(fit$details$fixedbeta)) {
        fb <- fit$details$fixedbeta
        fixed <- fb[max(unlist(fit$parindx)) + 1:2]
        pred[!is.na(fixed), 1] <- fixed[!is.na(fixed)]
    }
    if (angle == "degrees")
        pred["psiA", ] <- pred["psiA", ] * 360 / 2 / pi
    pred["psiR", ] <- 1 + exp(pred["psiR", ])  
    beta <- co["psiR","beta"]
    sebeta <- co["psiR","SE.beta"]
    pred["psiR", "SE.estimate"] <- exp(beta) * sqrt(exp(sebeta^2)-1)
    pred
}

anisotropic.fit <- function (..., psiA = pi/4, psiR = 2) {
    args <- list(...)
    if (is.null(args$details))
        args$details <- vector('list')
    args$details$userdist <- anisodistfn
    args$details$miscparm <- c(psiA = psiA, psiR = log(psiR-1))
    tmp <- do.call("secr.fit", args)
    tmp$call <- NULL ## drop bulky evaluated call
    tmp
}

R/simcapt.bvn.R
simcapt.bvn <- function (traps, popn, type = c('uniform', 'BVN'), g0 = 0.2, lambda0 = 0.2, p = 0.95,
                         noccasions = 5, renumber = TRUE) {
    
    traps.inside.ellipse <- function (traps, centre, theta, s2xy, p) {
        xy <- sweep (traps, MARGIN = 2, STATS = centre, FUN = '-')
        xy <- rotate (xy, -360/2/pi*theta)
        inside <- function (xy) {
            sum(xy^2/s2xy) < circular.r(p, hazard = FALSE)^2   ## 2.45 default
        }
        apply(xy,1,inside)
    }
    traps.pmvn <- function (traps, centre, theta, s2xy) {
        xy <- sweep (traps, MARGIN = 2, STATS = centre, FUN = '-')
        xy <- rotate (xy, -360/2/pi*theta)
        maxd <- 1/(2 * pi * prod(s2xy)^0.5)
        apply (xy, 1, mvtnorm::dmvnorm, mean = c(0,0),
               sigma = matrix(c(s2xy[1],0,0,s2xy[2]), nrow = 2)) / maxd
    }
    
    ## mainline
    if (!(detector(traps) %in% c('multi','proximity')) ){
        warning ("detectors other than multi or proximity will be treated as proximity detectors")
        detector(traps) <- 'proximity'
    }
    type <- match.arg(type)
    N <- nrow(popn)
    K <- nrow(traps)
    w <- array(0, dim = c(N,noccasions,K))
    s2xy <- attr(popn, 's2xy')
    theta <- attr(popn, 'theta')
    popn <- as.matrix(popn)
    for (i in 1:N) {
        if (type == 'uniform')  ## uniform ellipse
            OK <- traps.inside.ellipse( traps = traps, centre = popn[i,],
                                        theta = theta[i], s2xy = s2xy[i,], p = p )
        else                    ## BVN
            OK <- traps.pmvn( traps = traps, centre = popn[i,],
                              theta = theta[i], s2xy = s2xy[i,])
        for (s in 1:noccasions) {
            w[i,s,] <- OK   ## constant over occasions
        }
    }
    ## w remains numeric array dim = c(N,noccasions,K)
    if (detector(traps)=='proximity') {
        if (type == 'uniform') 
            g <- w * g0 ## 2018-07-26, for clarity
        else
            g <- 1 - exp(- w * lambda0)
        w <- 1 * (runif(prod(dim(w))) < g)  
        ## drop null capture histories
        caught <- apply(w,1,sum) > 0
        w <- w[caught,,,drop = FALSE]
    }
    else {
        # multi-catch traps
        # use hazard of detection
        if (type == 'BVN') 
            w <- w * lambda0
        else
            w <- w * -log(1-g0)
        sumhaz <- apply(w, 1:2, sum)
        w <- w / as.numeric(sumhaz)  # relative hazard
        w[is.na(w)] <- 0 # arises with uniform when zero overall Pr(capt)
        caughts <- (1 - exp(-sumhaz)) > runif(prod(dim(sumhaz))) 
        ## drop null capture histories
        caught <- apply(caughts,1,sum) > 0
        w <- w[caught,,,drop = FALSE]
        caughts <- caughts[caught,,drop = FALSE]
        # select one of traps with probability in w 
        # (w provides prob argument of sample.int)
        w <- caughts * apply(w, 1:2, sample.int, n=K, size=1, replace = FALSE)
    }
    
    n <- nrow(w)
    class(w) <- 'capthist'
    traps(w) <- traps
    session(w)           <- '1'           ## dummy session values for now
    if (renumber && (n>0)) rownames(w) <- 1:n
    else {
        rownames(w) <- rownames(popn)[caught]
    }
    w
}


R/subset.popn.R
## temporarily redefine subset.popn to respect s2xy, theta attributes (not needed)
subset.popn <- function (x, subset = NULL, sessions = NULL, poly = NULL,
                         poly.habitat = TRUE, keep.poly = TRUE, renumber = FALSE, ...)
    ## x - popn object
    ## subset - vector (character, logical or integer) to subscript rows (dim(1)) of x
    ## sessions - vector (integer or logical) to subscript sessions
{
    if (ms(x)) {
        if (!is.null(sessions))
            x <- x[sessions]
        out <- vector('list')
        for (i in 1:length(x)) {
            if (is.list(subset))
                sset <- subset[[i]]
            else
                sset <- subset
            if (is.null(subset))  ## 2015-03-17
                out[[i]] <- x[[i]]
            else
                out[[i]] <- subset(x[[i]], subset[[i]], NULL, renumber, ...)
        }
        class(out) <- c('list','popn')
        out
    }
    else {
        #-------------------------
        # default subset is all
        if (is.null(subset))
            subset <- 1:nrow(x)
        #-------------------------
        # restrict to a polygon
        # added 2011-10-20
        
        if (!is.null(poly)) {
            OK <- pointsInPolygon(x, poly)
            if (!poly.habitat)
                OK <- !OK
            subset <- subset[OK]
        }
        #-------------------------
        # apply subsetting
        pop <- x[subset,]
        #-------------------------
        if (renumber)
            rownames(pop) <- 1 : nrow(pop)
        
        class(pop) <- c('popn', 'data.frame')
        attr(pop, 'Ndist') <- NULL     ## no longer known
        attr(pop, 'model2D') <- NULL   ## no longer known
        attr(pop, 'boundingbox') <- attr(x, 'boundingbox')
        if (!is.null(poly) & keep.poly) {
            attr(pop, 'polygon') <- poly
            attr(pop, 'poly.habitat') <- poly.habitat
        }
        if (!is.null(covariates(x))) {
            covariates(pop) <- covariates(x)[subset,,drop=FALSE]
        }
        attr(pop, 's2xy') <- attr(x, 's2xy')[subset, ,drop = FALSE]
        attr(pop, 'theta') <- attr(x, 'theta')[subset]
        pop
    }
}


R/plotpopn.bvn.R
plotpopn.bvn <- function (popn, radius = 2.45, add = FALSE, region = NULL,  ...) {
    if (!is.null(region))
        requireNamespace('sp')
    if (!is.null(region) & !inherits(region, 'SpatialPolygons'))
        region <- SpatialPolygons(list(Polygons(list(Polygon(as.matrix(region))), ID=1)))
    N <- nrow(popn)
    s2xy <- attr(popn, 's2xy')
    theta <- attr(popn, 'theta')
    if (!add) plot(popn)
    popn <- as.matrix(popn)
    for (i in 1:N) {
        cov <- matrix(c(s2xy[i,1],0,0,s2xy[i,2]), ncol = 2)
        ellipsi <- ellipse(popn[i,], shape = cov, radius = radius)
        ellipsi <- rotate (ellipsi, theta[i]*360/2/pi, centrexy = popn[i,])
        ellipsesp <- SpatialPolygons(list(Polygons(list(Polygon(as.matrix(ellipsi))), ID=1)))
        if (!is.null(region)) {
            if (!requireNamespace("rgeos")) stop ("gIntersection uses rgeos that is unavailable")
            ellipsi2 <- gIntersection(ellipsesp, region)
            if (!is.null(ellipsi2))
                plot(ellipsi2, add = TRUE, ...)
            polygon(ellipsi)
        }
        else {
            polygon (ellipsi, ...)
        }
    }
}



R/runEllipseSim.R
## modified 2018-07-26 to also report predicted values 
## modified 2022-10-11 for random traps
runEllipseSim <- function(nrepl = 100, sigmaX = 25, sigmaY = 25, theta = NULL, 
    type = c('uniform', 'BVN'), g0 = 0.2, lambda0 = 0.4, p = 0.95, 
    traps = NULL, trapargs = NULL, noccasions = 5, buffer = 100, 
    D = 10, extractfn = derived, seed = NULL, ncores = NULL, 
    Ndist = 'fixed', secrfn = c('secr.fit', 'anisotropic.fit'),...) {
    onereplicate <- function (r) {

        if (is.function(traps)) {
            # function to deliver random trap locations
            # local 'traps' does not alter that argument of outer fn
            traps <- do.call(traps, trapargs)
            mask <- make.mask(traps, buffer = buffer, type = 'trapbuffer')
        }
    
        if (is.null(sigmaY) & !is.function(sigmaX)) {
            ## simple and direct circular using conventional secr::sim.capthist
            pop <- sim.popn(core = traps, buffer = buffer, D = D, Ndist = Ndist)
            CH <- sim.capthist(traps, pop, detectfngen, detectpar, noccasions)            
        }
        else {
            pop <- simpopn.bvn(s2xy = s2xy, theta = theta, core = traps, buffer = buffer, 
                               D = D, Ndist = Ndist)
            CH <- simcapt.bvn(traps, pop, type, g0, lambda0, p, noccasions)
        }
        out <- list(npop = nrow(pop), nCH = nrow(CH), ncapt=sum(CH), fit = nullfit)
        
        if (!is.null(secrfn)) {
            args <- list(...)
            args$capthist <- CH
            args$mask <- mask
            args$trace <- FALSE
            fit <- try(do.call(secrfn, args))
            if (!inherits(fit, "secr")) {
                out$fit <- nullfit
                out$pred <- nullpred
            }
            else {
                out$fit <- extractfn(fit)
                out$pred <- predict(fit)
            }
        }
        cat ('Completed ', r, date(), '\n')
        flush.console()
        out
    }
    secrfn <- match.arg(secrfn)
    if (is.null(traps)) {
        traps <- make.grid(detector = 'proximity')
    }
    else if (!is.function(traps)) {
        mask <- make.mask(traps, buffer = buffer, type = 'trapbuffer')
    }
    columns <- c('estimate','SE.estimate','lcl','ucl')
    
    type <- match.arg(type)
    nullfit <- as.data.frame(matrix(NA, nrow = 2, ncol = 4, 
                                    dimnames = list(c('esa','D'), columns)))
    nullpred <- as.data.frame(matrix(NA, nrow = 2, ncol = 5, 
                                    dimnames = list(c('lambda0','sigma'), c('link',columns))))
    
    ## for simcapt.bvn
    if (is.function(sigmaX))
        s2xy <- sigmaX
    else
        s2xy <- c(sigmaX, sigmaY)^2
    
    ## define detectpar etc. in case needed for sim.capthist
    detectfngen <- if (type=='uniform') 4 else 14
    if (is.null(sigmaY) & !is.function(sigmaX)) {
        if (detectfngen == 4)
            detectpar <- list(g0 = g0, sigma = sigmaX * circular.r(p, hazard = FALSE))
        else
            detectpar <- list(lambda0 = lambda0, sigma = sigmaX)
    }
    
    if (!is.null(ncores)) {
        setNumThreads(ncores)
    }
    
    set.seed(seed)        
    lapply(1:nrepl, onereplicate)
}